### PR TITLE
Write to stream

### LINKED ISSUE #492: Implement MCNP_Problem.write(file_handle)
**Is your feature request related to a problem? Please describe.**

In addition to `MCNP_Problem.write_to_file(new_problem)`, implement `MCNP_Problem.write_to_stream(file_stream)`. Then, users can write to any open file stream, including `io.StringIO`.

**Describe the solution you'd like**

```python
class MCNP_Problem:
    def write_to_file(self, str: file_path, bool: overwrite=False):
        """
        Check for existence
        Open the file handle
        Call self.write()`
        Close the file handle

        Also, rename kwarg from "new_problem" to "file_path" or something that
        does not imply that it is another ``MCNP_Problem`` instance.
        """
       . . .
```

and

```python
class MCNP_Problem:
    def write_to_stream(io.TextIOBase: file_handle):
        """
        Write to an open stream.

        Can allow to fail naturally, or do either of the following checks:
        >>> isinstance(file_handle, io.TextIOBase)
        >>> hasattr(file_handle, "write") and callable(getattr(file_handle, "write"))
        """
        . . .
```

**Describe alternatives you've considered**
The exact naming is up for discussion. I like `fpath` instead of `file_path` and `fhandle` or `fh` instead of `file_handle`.

Maybe `MCNP_Problem.write_to_stream()` should return something, such as the difference of `f.tell()` before and after writing in case the user wants to rewind the file (similar to a stream's `.write()`).

Another alternative is that `MCNP_Problem` could function more like a stream itself, with its own `.read()` and `.write()` methods. This seems unnecessary.

**Additional context**
The `.write_to_stream()` method would be highly useful in the MontePy integration tests.

### PR DESCRIPTION
**Pull request by tjlaboss**
# Description

* Implement `MCNP_Problem._write_to_stream()` 
* Implement `MCNP_Problem.write_problem()`, which handles both file paths and IO streams.
* Mark `MCNP_Problem.write_to_file()` with a deprecation warning.
* Update the integration and edge case tests accordingly.
* Include `montepy.errors` in top-level namespace
* Convert _test_integration.py_ from unittest to pytest, fixing some minor errors along the way.
* Update documentation accordingly.

Fixes #492

# Checklist

- [x] Either fully implement or deprecate `MCNP_InputFile`
- [x] Determine what `MCNP_Problem.write_*` methods should return
- [x] Settle on names for affected methods and parameters
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have updated the changelog
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: montepy/mcnp_problem.py [hunk 1]
```diff
@@ -390,25 +391,56 @@ def add_cell_children_to_problem(self):
         self._transforms = Transforms(transforms)
         self._data_inputs = sorted(set(self._data_inputs + materials + transforms))
 
-    def write_to_file(self, new_problem, overwrite=False):
+    def write_problem(self, destination, overwrite=False):
+        """
+        Write the problem to a file or writeable object.
+
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:397]: I'm not sure if this is the most pythonic name.

Comment (MicahGale) [reply montepy/mcnp_problem.py:397]: * Openpyxl: `save()` not a fan. This is very MS office centric
* pandas: `to_*()` e.g., `to_excel` `to_json`... I think this makes sense for something so filetype agnostic
* numpy: `tofile()`, `dump`
* openmc: `export_to_xml`
* json/pickle: `dump`

So nvm, there isn't a really "pythonic" way to do it. I think MontePy just needs to be consistent

So I propose:

1. `write` means to write out the state in a way that MCNP understands
2. `dump` means to serialize to a file (JSON or pickle)

`write` usually takes a string as an argument, so I think `write_problem` is a good name.

File: montepy/mcnp_problem.py [hunk 2]
```diff
@@ -390,25 +391,56 @@ def add_cell_children_to_problem(self):
...
         """
         Writes the problem to a file.
 
         .. versionchanged:: 0.3.0
             The overwrite parameter was added.
 
-        :param new_problem: the file name to write this problem to
-        :type new_problem: str
+        :param file_path: the file path to write this problem to
+        :type file_path: str
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:420]: Probably should add `os.pathlike`

File: montepy/mcnp_problem.py [hunk 3]
```diff
@@ -390,25 +391,56 @@ def add_cell_children_to_problem(self):
...
         :raises FileExistsError: if a file already exists with the same path.
         :raises IsADirectoryError: if the path given is actually a directory.
         """
-        new_file = MCNP_InputFile(new_problem, overwrite=overwrite)
-        with new_file.open("w") as fh, warnings.catch_warnings(
-            record=True
-        ) as warning_catch:
+        new_file = MCNP_InputFile(file_path, overwrite=overwrite)
+        with new_file.open("w") as fh:
+            self.write_to_stream(fh)
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:429]: I was about to freak about losing the warning_catch, but now I see what you did. I like it.

File: montepy/mcnp_problem.py [hunk 4]
```diff
@@ -390,25 +391,56 @@ def add_cell_children_to_problem(self):
...
+        # Consider returning result of write_to_stream()
+
+    def write_to_stream(self, fh):
+        """
+        Writes the problem to a writeable stream.
+
+        TODO: Expand MCNP_InputFile or adjust this method
+              to fully suppport generic file handles.
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:437]: I think that's the ultimate solution here. How it is getting the input file lineno is just `\n` counting, which we can totally do with a stream. 

As for `tell` I don't think it's necessary. `tell` has no correlation to line number (or even character count sometimes, and is only useful for `seek`. I don't think we should implement `seek`.

**Review by MicahGale — COMMENTED** [review]
Ok I like the interface for the users.

After reading everything I think the approach should be:

1. implement stream support for `MCNP_InputFile`
2. make `write_to_*` prepare an `MCNP_InputFile`
3. call a `_private` method with the InputFile in the right state that then handles the warnings.

File: montepy/mcnp_problem.py (hunk 1 above)
Comment (tjlaboss) [reply montepy/mcnp_problem.py:397]: That was my exact train of thought.

**Comment by tjlaboss** [comment]
The reported test failures on a56f2943522ba5a0df1534ba0287ddf22398f88c are from the deprecation warnings. All the unit tests actually pass.

I've implemented a new class method `MCNP_InputFile.from_open_stream()` that bypasses `MCNP_InputFile.open()`. Take a look and see what you think. If you like the functionality, let's think about the interface. In my opinion, `.__init__()` should actually be based on a stream. Then `.open()` could be converted to a `@classmethod` that creates the file handle and instantiates a new object.

**Comment by MicahGale** [comment]
TODO: update the test from #491 at the very least.

**Comment by MicahGale** [comment]
> The reported test failures on [a56f294](https://github.com/idaholab/MontePy/commit/a56f2943522ba5a0df1534ba0287ddf22398f88c) are from the deprecation warnings. All the unit tests actually pass.
> 
> I've implemented a new class method `MCNP_InputFile.from_open_stream()` that bypasses `MCNP_InputFile.open()`. Take a look and see what you think. If you like the functionality, let's think about the interface. In my opinion, `.__init__()` should actually be based on a stream. Then `.open()` could be converted to a `@classmethod` that creates the file handle and instantiates a new object.

Hmm interesting idea. I'm a bit hesitant this second. Look into how `read_input` works with this class first.

File: doc/source/changelog.rst [hunk 5]
```diff
@@ -1,13 +1,17 @@
 MontePy Changelog
 =================
 
-#Next Version#
---------------
-
+0.4.0
+-----
```
Comment (MicahGale) [inline doc/source/changelog.rst:5]: Not the time or place to change this.

File: doc/source/changelog.rst [hunk 6]
```diff
@@ -1,13 +1,17 @@
...
 **Bug Fixes**
 
 * Fixed bug that didn't show metastable states for pretty printing and isotope. Also handled the case that Am-241 metstable states break convention (:issue:`486`).
 
+**Features Added**
+
+* Write problems to either file paths or streams (file handles) with MCNP_Problem.write_problem() (:issue:`492`).
+
+
```
Comment (MicahGale) [inline doc/source/changelog.rst:14]: Features before bugs.

File: montepy/input_parser/input_file.py [hunk 7]
```diff
@@ -32,6 +32,19 @@ def __init__(self, path, parent_file=None, overwrite=False):
...
 
+    @classmethod
+    def from_open_stream(cls, fh):
+        """
+        Create an MCNP Input File from an open, writable stream
+
+        :param fh: An open and writable object, such as a file handle.
+        :type fh: io.TextIOBase
+        """
+        name = getattr(fh, "name", fh.__class__.__name__)
```
Comment (MicahGale) [inline montepy/input_parser/input_file.py:43]: Would `type(fh).__name__` be more pythonic?

File: montepy/mcnp_problem.py [hunk 8]
```diff
@@ -390,25 +391,58 @@ def add_cell_children_to_problem(self):
         self._transforms = Transforms(transforms)
         self._data_inputs = sorted(set(self._data_inputs + materials + transforms))
 
-    def write_to_file(self, new_problem, overwrite=False):
+    def write_problem(self, destination, overwrite=False):
+        """
+        Write the problem to a file or writeable object.
+
+        :param destination: File path or writable object
+        :type destination: {io.TextIOBase, str, os.PathLike}
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:399]: Is this proper sphinx?

File: montepy/mcnp_problem.py [hunk 9]
```diff
@@ -390,25 +391,58 @@ def add_cell_children_to_problem(self):
...
+                f"destination f{destination} is not a file path or writable object"
+            )
+
+    def write_to_file(self, file_path, overwrite=False):
         """
         Writes the problem to a file.
 
+        .. versionchanged:: 0.4.0
+            Deprecated in favor of self.write_problem()
+
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:421]: Should be a `.. deprecated::` then.

File: montepy/mcnp_problem.py [hunk 10]
```diff
@@ -390,25 +391,58 @@ def add_cell_children_to_problem(self):
...
         :raises FileExistsError: if a file already exists with the same path.
         :raises IsADirectoryError: if the path given is actually a directory.
         """
-        new_file = MCNP_InputFile(new_problem, overwrite=overwrite)
-        with new_file.open("w") as fh, warnings.catch_warnings(
-            record=True
-        ) as warning_catch:
+        cn = self.__class__.__name__
+        msg = f"{cn}.write_to_file() is deprecated. Use {cn}.write_problem()"
+        warnings.warn(msg, DeprecationWarning)
```
Comment (MicahGale) [inline montepy/mcnp_problem.py:435]: I'm not sure if we need to raise deprecation warnings yet.

File: tests/test_edge_cases.py [hunk 11]
```diff
@@ -41,13 +41,9 @@ def test_orphaning_mt(self):
             montepy.input_parser.block_type.BlockType.DATA,
         )
         problem.data_inputs.append(montepy.data_inputs.data_parser.parse_data(card))
-        try:
+        with io.StringIO() as stream:
             with self.assertRaises(MalformedInputError):
```
Comment (MicahGale) [inline tests/test_edge_cases.py:45]: `pytest.raises`

File: tests/test_integration.py
Comment (MicahGale) [inline tests/test_integration.py:1]: Did you add any new tests? TL;DR

**Review by MicahGale — CHANGES_REQUESTED** [review]

**Comment by MicahGale** [comment]
Coverage has gone down slightly.

File: doc/source/changelog.rst (hunk 5 above)
Comment (tjlaboss) [reply doc/source/changelog.rst:5]: Should I separate them into "#Next Minor Version#" and "#Next Patch#"?

File: montepy/input_parser/input_file.py (hunk 7 above)
Comment (tjlaboss) [reply montepy/input_parser/input_file.py:43]: I generally use `.__class__.__name__` for printing, to respect an overwrite of `.__class__` if that should occur, but have no strong feelings on the matter.

File: montepy/mcnp_problem.py (hunk 10 above)
Comment (tjlaboss) [reply montepy/mcnp_problem.py:435]: `PendingDeprecationWarning`, or no warning at all?

File: tests/test_integration.py
Comment (tjlaboss) [reply tests/test_integration.py:1]: The existing tests have been moved to use `write_problem()` in all but one case, which continues to test `write_to_file()` with `pytest.warns(DeprecationWarning)`.

`test_importance_rewrite()` tests `write_problem()` with two real files. The rest test `write_problem()` with a stream, specifically a `StringIO`.

File: tests/test_integration.py [hunk 12]
```diff
@@ -1,1048 +1,1069 @@
...
-        surf.transform.number = 5
-        self.assertEqual(int(surf.format_for_mcnp_input((6, 2, 0))[0].split()[1]), 5)
-        # test changing surface constants
-        new_prob = copy.deepcopy(problem)
-        surf = new_prob.surfaces[4]
-        surf.location = 2.5
-        self.assertEqual(
-            float(surf.format_for_mcnp_input((6, 2, 0))[0].split()[-1]), 2.5
-        )
+@pytest.fixture
```
Comment (MicahGale) [inline tests/test_integration.py:26]: You should make this have a scope so it's ran once. Not being able to cache these problems really slowed down the test suite for a while. This is what I came up with recently: https://github.com/idaholab/MontePy/pull/488/files#diff-4071f2914f1a922463edf531fe1ccb42d1f0534ecb7f22b7812d955c840f1f46R295-R304

File: doc/source/changelog.rst (hunk 5 above)
Comment (MicahGale) [reply doc/source/changelog.rst:5]: No the plan is to do one big release soon

Comment (MicahGale) [reply doc/source/changelog.rst:5]: #Next Version# is used by `grep`

File: montepy/input_parser/input_file.py (hunk 7 above)
Comment (MicahGale) [reply montepy/input_parser/input_file.py:43]: Ok that's good enough for me.

File: tests/test_integration.py
Comment (tjlaboss) [reply tests/test_integration.py:1]: I just added a new test that ensures that `write_problem()` catches an unwritable object.

File: montepy/mcnp_problem.py (hunk 10 above)
Comment (MicahGale) [reply montepy/mcnp_problem.py:435]: pending is an option? I'm trying to think: what would @bascandr say?

I honestly don't see any reason to truly deprecate it and remove it. I think no warning. I'm fine with users not making changes at this point.

File: tests/test_integration.py
Comment (MicahGale) [reply tests/test_integration.py:1]: I think you'll need to write a few more tests to ensure 100% coverage of the changes.

File: montepy/mcnp_problem.py (hunk 9 above)
Comment (tjlaboss) [reply montepy/mcnp_problem.py:421]: undid deprecation

File: tests/test_integration.py
Comment (tjlaboss) [reply tests/test_integration.py:1]: All new code is now covered

File: tests/test_integration.py (hunk 12 above)
Comment (tjlaboss) [reply tests/test_integration.py]: Set the scope with [5460cb2](https://github.com/idaholab/MontePy/pull/493/commits/5460cb22d0ba8fcc63a88c5e667a223327196671)

After all the tests have been converted to pytest, we should consider making a global fixtures module. `test.imcnp` gets used a lot, for instance.

**Comment by tjlaboss** [comment]
Re-ady for re-review.

**Comment by MicahGale** [comment]
> Re-ady for re-review.

image
There's a button for that.

**Comment by tjlaboss** [comment]
the button doesn't come with a quirky message tho

**Comment by MicahGale** [comment]
Touche.

**Review by MicahGale — APPROVED** [review]
LGTM. 

In the feature I can't decide if most test cases should move to `IO.StringIO`.
